File: python/sampleClient.py
import logging
import grpc
import time
import base64

from google.protobuf.timestamp_pb2 import Timestamp

import TrafikVerketExample_pb2
import TrafikVerketExample_pb2_grpc

def run():
    with grpc.insecure_channel('127.0.0.1:50051') as channel:
        stub = TrafikVerketExample_pb2_grpc.WeatherServiceStub(channel)
        req = TrafikVerketExample_pb2.WeatherMeasurepointRequest(ids = [1,2])

        camReq = TrafikVerketExample_pb2.CameraRequest(id = 1)

        resp = stub.GetCameraData(camReq)
        image_string = base64.b64decode(resp.image)
        fh = open("imageToSave.png", "wb")
        fh.write(image_string)
        fh.close()

        print("got image " + str(resp.image))

        timestamp = Timestamp()
        wmpList = []
        wmp1 = TrafikVerketExample_pb2.WeatherMeasurepoint(id=1, name="station01", Observation_Air_Temperature=20.0, last_update=timestamp.GetCurrentTime())
        wmp2 = TrafikVerketExample_pb2.WeatherMeasurepoint(id=2, name="station02", Observation_Air_Temperature=20.1, last_update=timestamp.GetCurrentTime())
        wmpList.append(wmp1)
        wmpList.append(wmp2)

        resp = stub.AddWeatherMeasurepoint(TrafikVerketExample_pb2.WeatherMeasurepointList(points = wmpList))
        print("added measurement points " + str(resp.successful))
        
        for i in range(5):
            resp = stub.GetWeatherMeasurepoint(req)
            print(resp)
            time.sleep(0.8)            

if __name__ == '__main__':
    logging.basicConfig()
    run()

File: cpp/src/sampleClient.cpp
#include <iostream>

#include <grpc/grpc.h>
#include <grpcpp/channel.h>
#include <grpcpp/create_channel.h>
#include <grpcpp/client_context.h>

#include "google/protobuf/util/time_util.h"

#include "sampleClient.hpp"

#include "../build/TrafikVerketExample.pb.h"
#include "../build/TrafikVerketExample.grpc.pb.h"

using namespace std;
using namespace sampleclient;
using namespace trafikverket;

using google::protobuf::util::TimeUtil;

void MyClient::callAPI() {
    cout << "start gRPC call " << std::endl;

    string host = "localhost:50051";
    auto creds = grpc::InsecureChannelCredentials();
    auto channel = grpc::CreateChannel(host, creds);
    std::unique_ptr<trafikverket::WeatherService::Stub> stub = WeatherService::NewStub(channel);

    grpc::ClientContext context;

    WeatherMeasurepointRequest req;
    WeatherMeasurepointList resp;
    stub->GetWeatherMeasurepoint(&context, req, &resp);
    auto points = resp.points();
    for(::google::protobuf::RepeatedPtrField<WeatherMeasurepoint>::const_iterator it = points.begin() ; it != points.end(); ++it ) {
        string data = "";
        data += std::to_string(it->id());
        data += " " + it->name();
        data += " " + std::to_string(it->observation_air_temperature());
        data += " " +  TimeUtil::ToString(it->last_update());
        cout << data << std::endl;
    }
    

}